statsprinter.go

```diff
@@ -255,95 +256,174 @@ JSON output section
 
 */
 
-/* Print message  in JSON format */
-func jsonPrintf(format string, a ...interface{}) {
-	data := struct {
-		Message string `json:"message"`
-	}{
-		Message: fmt.Sprintf(format, a...),
-	}
-	outputJson, _ := json.Marshal(&data)
-	fmt.Println(string(outputJson))
+type JSONEventType string
+
+const (
+	// Probe is an event type that represent 1 probe / ping / request event.
+	Probe JSONEventType = "probe"
+	// Retry is an event type that's being sent,
+	// when tcping retries to resolve a hostname.
+	Retry JSONEventType = "retry"
+	// RetrySuccess is a sub-type for of Retry event,
+	// when previous retry was unsuccessful, but the next one became successful.
+	RetrySuccess JSONEventType = "retry-success"
+	// Start is an event type that's sent only one,
+	// before doing any actual work.
+	Start JSONEventType = "start"
+	// Stats is a final event that's being sent
+	// before tcping exits.
+	Stats JSONEventType = "stats"
+)
+
+// jsonEncoder stores the encoder for json output.
+// It could be used to tweak default options, like indentation or
+// change output to another writer interface.
+var jsonEncoder = json.NewEncoder(os.Stdout)
+
+// printJson is a shortcut for Encode() on jsonEncoder.
+var printJson = jsonEncoder.Encode
+
+// JSONData contains all possible fields for JSON output.
+// Because one event usually contains only a subset of fields,
+// other fields will be omitted in the output.
+type JSONData struct {
+	// Type is a mandatory field that specifies type of a message/event.
+	// Possible types are:
+	//	- start
+	// 	- probe
+	// 	- retry
+	// 	- stats
+	//	- retry-success
+	Type JSONEventType `json:"type"`
+	// Message contains a human-readable message.
+	Message string `json:"message"`
+	// Timestamp contains data when a message was sent.
+	Timestamp time.Time `json:"timestamp"`
+
+	// Optional fields below
+
+	Addr                 string `json:"addr,omitempty"`
+	Hostname             string `json:"hostname,omitempty"`
+	HostnameResolveTries uint   `json:"hostname_resolve_tries,omitempty"`
+	IsIP                 *bool  `json:"is_ip,omitempty"`
+	Port                 uint16 `json:"port,omitempty"`
```

Comment:
Any particular reason for exporting these fields (and the rest)?

Reply:
Yes, because otherwise they won't be in the output. Go marshals/unmarshals only exported fields. [go playground example](https://go.dev/play/p/dNiaIBUD16M)
If you want to go in-depth, there's an [official article](https://go.dev/blog/json) on this, or a [simple SO answer](https://stackoverflow.com/questions/11126793/json-and-dealing-with-unexported-fields) on this.

Reply:
Well, technically, we can make them unexported and write a custom marshaller, but I think that's a huge overkill for this case.

Reply:
ohhh I see. did not know about that.
Thanks a lot.

It is fine to keep them as they are then.